File: rag-server/src/indexBuilder.js
"use strict";

const fs = require("fs/promises");
const path = require("path");
const { embedText } = require("./geminiClient");

const MAX_FILE_CHARACTERS = 20000;
const CHUNK_SIZE = 1200;
const CHUNK_OVERLAP = 160;
const ENABLE_EMBEDDINGS = String(process.env.RAG_ENABLE_EMBEDDINGS || "false") === "true";

const TARGETS = [
  "README.md",
  "ProjectScreen.md",
  "SiteScreen.md",
  "webapp/manifest.json",
  "webapp/i18n/i18n.properties",
  "webapp/controller",
  "webapp/view",
  "webapp/utils"
];

const ALLOWED_EXTENSIONS = new Set([".md", ".xml", ".js", ".properties", ".json"]);

async function readExistingFile(filePath) {
  try {
    const stat = await fs.stat(filePath);
    return stat.isFile();
  } catch {
    return false;
  }
}

async function collectFiles(baseDir) {
  const files = [];

  for (const relativeTarget of TARGETS) {
    const absoluteTarget = path.join(baseDir, relativeTarget);
    if (!(await readExistingFile(absoluteTarget))) {
      await collectRecursively(absoluteTarget, files);
      continue;
    }

    files.push(absoluteTarget);
  }

  return files;
}

async function collectRecursively(entryPath, files) {
  try {
    const stat = await fs.stat(entryPath);

    if (stat.isFile()) {
      const ext = path.extname(entryPath).toLowerCase();
      if (ALLOWED_EXTENSIONS.has(ext)) {
        files.push(entryPath);
      }
      return;
    }

    if (!stat.isDirectory()) {
      return;
    }

    const entries = await fs.readdir(entryPath, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name.startsWith(".")) {
        continue;
      }

      const absoluteChild = path.join(entryPath, entry.name);
      if (entry.isDirectory()) {
        await collectRecursively(absoluteChild, files);
      } else {
        const ext = path.extname(entry.name).toLowerCase();
        if (ALLOWED_EXTENSIONS.has(ext)) {
          files.push(absoluteChild);
        }
      }
    }
  } catch {
    // Ignore missing paths so this script can run across environments.
  }
}

function normalizeText(rawText) {
  return rawText
    .replace(/\r/g, "")
    .replace(/\t/g, "    ")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}

function splitIntoChunks(text) {
  if (!text) {
    return [];
  }

  const chunks = [];
  let start = 0;

  while (start < text.length) {
    const end = Math.min(start + CHUNK_SIZE, text.length);
    const chunkText = text.slice(start, end).trim();

    if (chunkText) {
      chunks.push(chunkText);
    }

    if (end >= text.length) {
      break;
    }

    start = Math.max(0, end - CHUNK_OVERLAP);
  }

  return chunks;
}

async function buildKnowledgeIndex(baseDir, outputPath) {
  const files = await collectFiles(baseDir);
  const chunks = [];
  let embeddingEnabled = ENABLE_EMBEDDINGS;
  let embeddedChunkCount = 0;
  let embeddingsDisabledReason = null;

  for (const absoluteFilePath of files) {
    const relativePath = path.relative(baseDir, absoluteFilePath).replace(/\\/g, "/");
    const rawText = await fs.readFile(absoluteFilePath, "utf8");
    const normalized = normalizeText(rawText).slice(0, MAX_FILE_CHARACTERS);
    const fileChunks = splitIntoChunks(normalized);

    for (let index = 0; index < fileChunks.length; index += 1) {
      const text = fileChunks[index];
      let embedding = null;

      if (embeddingEnabled) {
        try {
          embedding = await embedText(text);
          embeddedChunkCount += 1;
        } catch (error) {
          embeddingEnabled = false;
          embeddingsDisabledReason = error.message;
        }
      }

      chunks.push({
        id: `${relativePath}#${index + 1}`,
        source: relativePath,
        text,
        embedding
      });
    }
  }

  const indexData = {
    generatedAt: new Date().toISOString(),
    chunks
  };

  await fs.mkdir(path.dirname(outputPath), { recursive: true });
  await fs.writeFile(outputPath, JSON.stringify(indexData, null, 2), "utf8");

  return {
    fileCount: files.length,
    chunkCount: chunks.length,
    embeddingEnabled,
    embeddedChunkCount,
    embeddingsDisabledReason,
    outputPath
  };
}

module.exports = {
  buildKnowledgeIndex
};

File: rag-server-py/index_builder.py
"""Knowledge index builder for BuildTrack RAG."""

from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List

from gemini_client import embed_text

MAX_FILE_CHARACTERS = 20000
CHUNK_SIZE = 1200
CHUNK_OVERLAP = 160
ENABLE_EMBEDDINGS = os.getenv("RAG_ENABLE_EMBEDDINGS", "false").lower() == "true"

TARGETS = [
    "README.md",
    "ProjectScreen.md",
    "SiteScreen.md",
    "webapp/manifest.json",
    "webapp/i18n/i18n.properties",
    "webapp/controller",
    "webapp/view",
    "webapp/utils",
]

ALLOWED_EXTENSIONS = {".md", ".xml", ".js", ".properties", ".json"}


def collect_recursively(entry_path: Path, files: List[Path]) -> None:
    try:
        if entry_path.is_file():
            if entry_path.suffix.lower() in ALLOWED_EXTENSIONS:
                files.append(entry_path)
            return

        if not entry_path.is_dir():
            return

        for child in entry_path.iterdir():
            if child.name.startswith("."):
                continue
            if child.is_dir():
                collect_recursively(child, files)
            elif child.suffix.lower() in ALLOWED_EXTENSIONS:
                files.append(child)
    except FileNotFoundError:
        # Ignore missing paths so this script can run across environments.
        return


def collect_files(base_dir: Path) -> List[Path]:
    files: List[Path] = []

    for relative_target in TARGETS:
        absolute_target = base_dir / relative_target
        if absolute_target.is_file():
            files.append(absolute_target)
            continue
        collect_recursively(absolute_target, files)

    return files


def normalize_text(raw_text: str) -> str:
    return (
        raw_text.replace("\r", "")
        .replace("\t", "    ")
        .replace("\n\n\n", "\n\n")
        .strip()
    )


def split_into_chunks(text: str) -> List[str]:
    if not text:
        return []

    chunks: List[str] = []
    start = 0
    length = len(text)

    while start < length:
        end = min(start + CHUNK_SIZE, length)
        chunk_text = text[start:end].strip()
        if chunk_text:
            chunks.append(chunk_text)

        if end >= length:
            break

        start = max(0, end - CHUNK_OVERLAP)

    return chunks


def build_knowledge_index(base_dir: Path, output_path: Path) -> Dict[str, object]:
    files = collect_files(base_dir)
    chunks: List[Dict[str, object]] = []

    embedding_enabled = ENABLE_EMBEDDINGS
    embedded_chunk_count = 0
    embeddings_disabled_reason = None

    for absolute_file_path in files:
        relative_path = absolute_file_path.relative_to(base_dir).as_posix()
        raw_text = absolute_file_path.read_text(encoding="utf-8")
        normalized = normalize_text(raw_text)[:MAX_FILE_CHARACTERS]
        file_chunks = split_into_chunks(normalized)

        for index, text in enumerate(file_chunks, start=1):
            embedding = None
            if embedding_enabled:
                try:
                    embedding = embed_text(text)
                    embedded_chunk_count += 1
                except Exception as error:  # pylint: disable=broad-except
                    embedding_enabled = False
                    embeddings_disabled_reason = str(error)

            chunks.append(
                {
                    "id": f"{relative_path}#{index}",
                    "source": relative_path,
                    "text": text,
                    "embedding": embedding,
                }
            )

    generated_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    index_data = {
        "generatedAt": generated_at,
        "chunks": chunks,
    }

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(json.dumps(index_data, ensure_ascii=True, indent=2), encoding="utf-8")

    return {
        "fileCount": len(files),
        "chunkCount": len(chunks),
        "embeddingEnabled": embedding_enabled,
        "embeddedChunkCount": embedded_chunk_count,
        "embeddingsDisabledReason": embeddings_disabled_reason,
        "outputPath": str(output_path),
    }
